networkx/algorithms/isomorphism/tree_isomorphism.py

```diff
@@ -1,279 +1,686 @@
+        of v in the rooted tree. This map determines who are the parents of the
+        leaves found on the 0-th level.
+
+    PARENTHOOD : dict of node: node
+        The parenthood map of the rooted tree.
+
+    height : int
+        The height of the rooted tree.
+
+    Note
+    ----
+    This algorithm runs in O(n) time and space.
 
-    # our new "fake" root node is 0
-    # t1 is numbers from 1 ... n
-    # t2 is numbered from n+1 to 2n
     """
+    # Define an empty map for LEVELS, VALUES, CHILDREN and PARENTHOOD.
+    LEVELS = defaultdict(set)
+    CHILDREN = defaultdict(list)
+    VALUES = {}
+    PARENTHOOD = {}
 
-    dT = nx.DiGraph()
+    # Auxiliary variables.
+    from_distance = defaultdict(set)
+    distance_to_root = {}
+    distance_to_root[root] = 0
+    from_distance[0].add(root)
+    max_distance_to_root = 0
 
-    newroot1 = 1  # left root will be 1
-    newroot2 = nx.number_of_nodes(t1) + 1  # right will be n+1
+    # Perform a single BFS traversal to obtain the necessary information.
+    for (parent, child) in nx.bfs_edges(T, root):
+        # Define child's parent.
+        PARENTHOOD[child] = parent
 
-    # may be overlap in node names here so need separate maps
-    # given the old name, what is the new
-    namemap1 = {root1: newroot1}
-    namemap2 = {root2: newroot2}
+        # The parent can't be a leave, assign its value to -1.
+        VALUES[parent] = -1
 
-    # add an edge from our new root to root1 and root2
-    dT.add_edge(0, namemap1[root1])
-    dT.add_edge(0, namemap2[root2])
+        # The child could be a leave, assing its value to 0.
+        VALUES[child] = 0
 
-    for i, (v1, v2) in enumerate(nx.bfs_edges(t1, root1)):
-        namemap1[v2] = i + namemap1[root1] + 1
-        dT.add_edge(namemap1[v1], namemap1[v2])
+        # Define d(child, root) = d(parent, root) + 1
+        distance_to_root[child] = distance_to_root[parent] + 1
 
-    for i, (v1, v2) in enumerate(nx.bfs_edges(t2, root2)):
-        namemap2[v2] = i + namemap2[root2] + 1
-        dT.add_edge(namemap2[v1], namemap2[v2])
+        # Add the child to the set of vertices have the same distance to the
+        # root.
+        d_child = distance_to_root[child]
+        from_distance[d_child].add(child)
 
-    # now we really want the inverse of namemap1 and namemap2
-    # giving the old name given the new
-    # since the values of namemap1 and namemap2 are unique
-    # there won't be collisions
-    namemap = {}
-    for old, new in namemap1.items():
-        namemap[new] = old
-    for old, new in namemap2.items():
-        namemap[new] = old
+        # If the previous distance is greater than the max, update it.
+        if max_distance_to_root < distance_to_root[child]:
+            max_distance_to_root = distance_to_root[child]
 
-    return (dT, namemap, newroot1, newroot2)
+    # The max distance to the root is the rooted tree's height.
+    height = max_distance_to_root
 
+    # Build the LEVELS map.
```

Comment:
maybe slightly more pythonic : `LEVELS =  {height-d:s for (d, s) in from_distance.items() }` .